### ISSUE TITLE
VR IN POTREE???

### ISSUE THREAD

**Issue by mihapajnic** [potree/potree#801]
Hello,

is there any tutorial where VR would be enabled in Potree?

**Comment by midnight-dev** [comment]
Unfortunately, this isn't a simple answer. The WebVR spec, which Potree tries to make use of, was never ratified, and is now considered obsolete. WebXR is supposed to be its replacement - attempting to support both VR and AR with a single web API. _However_, WebXR is still in an early (but _working_) draft phase, and significant portions are not implemented in most browsers, which is expected for an evolving standard.

Right, so Potree wants to use of the WebVR API, which still lives on, kind of. It **DOES NOT** exist in Chrome - they previously had it behind a special flag, but it's been removed since some time last year. It **DOES** exist in the current version of Firefox with some caveats. You need to run it on Windows, and it **must** be connected to the website via HTTPS. That means you cannot use the official website [potree.org](http://potree.org), nor a [localhost server](http://potree.org/getting_started.html) to test this, as they do not support secured connections. (The official website might fix this in the future though... Hopefully.) 

Firefox is serious about that secured connection requirement. I thought `navigator.getVRDisplays` didn't exist on Firefox for a minute there, but that was because I was using _localhost_. The API just won't be available to JavaScript at all unless it's HTTPS, and currently, Potree tests for VR capabilities by using `getVRDisplays`.

I don't have a VR headset on hand, so I don't know how it should behave, but in theory, if you're accessing the current version of Potree through a **secure** website, using Firefox, and your VR headset is attached & running, then Potree should render in VR. If it doesn't, you may have to give it a push with something like `await viewer.startVR()`.

For reference, here's the portion of the render loop that determines if it's going to render in VR or not:

```javascript
const vrActive = (this.vr && this.vr.display.isPresenting);

if(vrActive){
	const {display, frameData} = this.vr;

	display.requestAnimationFrame(this.loop.bind(this));

	display.getFrameData(frameData);

	this.update(this.clock.getDelta(), timestamp);

	this.render();

	this.vr.display.submitFrame();
}else{
	requestAnimationFrame(this.loop.bind(this));

	this.update(this.clock.getDelta(), timestamp);

	this.render();
}
```

@potree / @m-schuetz Markus could give you better information regarding this, since he's the author and maintainer. I can tell from the commit history that this was the initial stepping stone to getting VR functionality in Potree, but I'm not sure if this code works or if there are plans to adapt & expand it. Or maybe change it entirely? No clue 🤷‍♂️

**Comment by mihapajnic** [comment]
Hello,

thank you very much for the comprehensive answer !! I know that said matter is not easy. Do you orda know any tools where this could be done easily. For now, I know that it can be done in FAROSCENE, but it's payable, so I'm looking for free solutions.

thanks for the reply

**Comment by m-schuetz** [comment]
VR support in Potree is a bit of a mess right now because support in browsers is quite unstable. VR using WebVR used to work fine in Firefox until about half a year ago, then things broke down. I think they even require https hosted content for VR now, which is messy because Potree is hosted on non-https sites. This also makes it hard to host pages on one site and data on another. 

WebVR is supposed to be deprecated and replaced by WebXR, but I've never gotten any WebXR page to work so far, so I can't develop WebXR support.

**Comment by jo-chemla** [comment]
We, at Iconem, did manage to get VR with potree working both on Firefox on Windows with a Vive headset, and using an Mozilla VR on an Oculus Quest. We indeed needed to use https on our remote server to access device orientation. 

Looking at this official [WebVR to WebXR migration guide](https://github.com/immersive-web/webxr/blob/master/webvr-migration.md), there are recommandations to trick the secure origin need. Among the listed solutions, two seem easy to use: 
 - first localhost should be regarded as a secure origin, 
 - otherwise, you can run chrome with the `--unsafely-treat-insecure-origin-as-secure="http://example.com"` flag. To do so, either close any instance of chrome, or add this tag as well: ` --user-data-dir="C:/Chrome_tmp_session" `

**Comment by mihapajnic** [comment]
Hello, thank you very much for your help. Now happy to work, I hope you manage to do a demo. Thanks again for the help !! LP

**Comment by midnight-dev** [comment]
> Hello,
> 
> thank you very much for the comprehensive answer !! I know that said matter is not easy. Do you orda know any tools where this could be done easily. For now, I know that it can be done in FAROSCENE, but it's payable, so I'm looking for free solutions.
> 
> thanks for the reply

If you're using for viewer for any commercial services, you could charge clients an additional fee if they need VR features, and that would cover the cost of expensive solutions like Faro. Of course, it could cost more than customers are willing to pay. I haven't looked.

If it's for purely personal or academic research purposes, it's harder to justify spending money - especially for subscription services. I get it. There's one other platform I'm quite familiar with and that's CesiumJS, which supports a VR mode. They offer world-scale rendering, and it's legitimately impressive to see what they've done with JavaScript, WebGL shaders, and worker threads. They offer _some_ support for point clouds, but only if they're converted to 3D Tiles, and you'll likely have to upload to their proprietary tiler, which involves signing up for their API. It **cannot** be used for commercial uses unless you pay out something like $150 per month, so I'd caution careful consideration. By the way, their JavaScript library is powerful, but the learning curve is steep. There are so many classes and methods and settings that it can get overwhelming. Thankfully, there's plenty of documentation. It's worth a peek, but learning about it, applying for their server API, and using it _just_ for a VR mode would be downright silly. 

If you can be patient, VR will come to Potree. I can't promise it will make it to the official repo, but I'll be working on a WebXR solution (plus a fallback pseudo VR mode) on my fork over the coming weeks. Fair warning: it may be a little jank because I don't have a VR headset. Even still, it's officially on my to-do list. 👌

**Comment by rooi** [comment]
Hi, I did some hacking to try to get VR working again using WebXR to, hopefully, go towards an AR iOS example/proof of concept. [https://github.com/MozillaReality/webxr-ios-js](url) 

The VR button is working and I can see the right eye image. The left is not correct, perhaps due to the incorrect projectionMatrix and viewTransform in the onXRFrame method.
[https://github.com/rooi/potree](url)  

Any pointers on how to get it correct?

It seems the AR iOS example will be to much work, for myself right now. It would be very interesting though seeing how normal potree looks on my iPhone.

**Comment by midnight-dev** [comment]
@rooi Your last commit in [rooi/potree](https://github.com/rooi/potree)

> **20** changed files with **14,693** additions and **11,558** deletions.

You're killing me, mate. 🌵 
I think I've found the relevant changes though. I don't have an iOS device, so I can't test specifically for Apple features, but the WebXR emulator does pick up the right frame like you mentioned. Rather, I can see an image on the right side of my screen. I can't confirm if the emu is actually getting any frames because that doesn't seem to be a feature of it.

When you mentioned going towards AR iOS, does that mean you followed an AR WebXR example? If so, that would explain why there's only on eye working since most AR right now projects to a single screen. I'll be giving this a poke today, but right now, all I can confirm is you did manage to get WebXR on the right side, which is progress 👍

**Comment by rooi** [comment]
oeps, sorry about that. Must be the threejs update (I picked the latest dev version). The other files are better distinguishable.

I mostly followed the WebVR to WebXR migration as you mentioned. First trying to fix VR before diving into an AR guide. Note I'm no web developer so be mindful for strange code ;)

**Comment by midnight-dev** [comment]
> Note I'm no web developer so be mindful for strange code ;)

Don't sell yourself short. You did well. Since this is an emerging API, we're all rookies here.

Alright, so I got it sorted, kinda. It works if you have a wildly uneven face with one eye 10 meters above/below the other. But, there are technically two points of view and 60 frames per second. Success?

![image](https://user-images.githubusercontent.com/47105073/81481377-522e4300-91f5-11ea-90f3-97b5de59fd8f.png)

I'm less than thrilled with my own jank, and I need to fix the dramatically different angles before I commit anything. And unfortunately, I have to work on some CloudFront nonsense today. But I'll fix it up for public consumption as soon as I can.

By the way, I'd like to have WebVR and a fallback method as well for browsers that don't support WebVR or WebXR. These changes cannibalize the existing WebVR code from 8 months ago, so I'll still be working on a solution to have all methods available instead of just one.

**Comment by rooi** [comment]
Ah getting close! I hope the last bits don't give you too much trouble.

A semi-related question since I'm trying to work towards iOS, if you don't mind. Do you know what could cause the pointclouds not to be shown on iOS while they are fine on pc? I've seen mentions and error related to GL_EXT_frag_depth not being supported and I've removed all things related to this (breaking EDL functionally, but just as an experiment). All warnings/erros are gone. EDL disabled works find on both iOS and PC. Still enabling EDL only shows the background with the model on iOS while pc still shows it.

Any idea what could cause this or how to debug it? Could it be something trivial as near/far planes and/or buffers that differ between iOS and PC?

**Comment by rooi** [comment]
@midnight-dev WebXR should provide a fallback for webVR right?
https://github.com/immersive-web/webxr-polyfill

**Comment by midnight-dev** [comment]
> @midnight-dev WebXR should provide a fallback for webVR right?

In the long-term, yes @rooi, but there are a few issues keeping it from being a good replacement candidate - for me, anyway. 

One thing is that VR and AR are only marginally related to each other. Requiring a hefty amount of AR code just the support the WebXR spec/polyfill that would _then_ attempt to fulfill the VR of WebVR seems like the wrong way to handle this. 

Secondly, it looks like their frame data mechanisms are not as related as I would have hoped. This means that trying to emulate and code for one will not be compatible with the other. Not by default. They're two different APIs that were developed years apart from one another. While they both share some of the same goals, neither was made _for_ the other. That's one of the reasons why you had to copy in starter snippets for WebXR and completely disable WebVR. It would've been nice if you could just include the polyfill, add the `xrCompatible` flag to the context, and then have WebVR through WebXR, but the two are just not compatible in that way. WebXR is still new enough that the implementation across the major browsers is far from final. Browsers may change how the interface is exposed or they could limit problematic features. There's a lot promise, but it's going to take time to see how things shake out. 

_That_ is why I want to keep WebVR, which has been in Firefox for years (and in Chrome behind a flag). It would be ideal to have a few lines of code that support it until WebXR matures enough for Firefox to obsolete WebVR. Even then, there should be a fallback for those who don't have a VR headset but can still enjoy VR content on their phone or by aligning their eyes & shifting focus. WebXR doesn't run when _not_ using a connected headset; nor does WebVR for that matter, which I guess makes sense. But that's why I'm calling it a fallback. That, in particular, will be on a separate branch for a while. I'll get the WebXR port committed sooner than the WebVR & fallback compatibility.

And yes, I'm still working on this. Just had a minor deviation in plans 🌵 

I decided to put some work into the mobile/iOS problem with Eye Dome Lighting. I'm not 100% sure I can make that work, but check it out my rambling if you're curious :tada:  https://github.com/potree/potree/issues/716#issuecomment-628245414

**Comment by rooi** [comment]
@midnight-dev the iOS EDL problem is interesting as well.
I gave webXR another go and stereo is working In my webXR branch. It does not respond to head movement yet, but I've got a local change which responds and just has to be implemented correctly.

**Comment by rooi** [comment]
The polyfill seems to work quite ok. I've just hacked together an AR version on iOS. The orientation is not perfect, Euler and no roll, but ik looks ok:
![IMG_0302](https://user-images.githubusercontent.com/5172371/83353977-89a49100-a356-11ea-98a5-08e1b47c7e2c.gif)

**Comment by m-schuetz** [comment]
@rooi This looks awesome!

**Comment by mihapajnic** [comment]
Hello,

@rooi
or maybe some VR option is already working without any pre-knowledge of the javascript programming language?

Thank you very much for the answers

**Comment by rooi** [comment]
@mihapajnic I've tested the webXR branch in my fork using the VR emulator in Chrome but not on a real headset. The emulator show both left and right eye images and it responds to movement of the emulated VR device.

The image above is AR on iOS which also works but has a problem with orientation/quaternions which I did not fix. Probably the same problem will be apparent on a real VR headset as well.

You could try the webXR branch and see what it brings. I suspect that it will not work fully and some javascript programming needs to be done to get it to work correctly

**Comment by renoir** [comment]
I know this is a dead thread but just to throw in my experiments:

I've managed to get potree / point clouds to display in proper VR using a mixture of stuff.
Starting with AFrame, get an simple example running, hosted locally on HTTPS server (I used the default Python HTTPS server)
Run it in Firefox, with a VR headset attached (I have a Valve Index) and the VR button works and goes into VR.

There's an aframe potree loader ( https://github.com/mattrei/aframe-potree-loader-component ). It uses potree1.7 which someone has extracted the loader functionality into https://github.com/pnext/three-loader. 
The versions of the loader and three.js need to be updated and it needs rebuilding, but it works.
It's easier to play with the HTML and use a normal browser to get the point cloud displayed.

This bodge allows me to show the example lion in full VR.
![screencapture-localhost-4443-cgi-vehiclepc-py-2021-06-18-09_56_18](https://user-images.githubusercontent.com/1187365/122535649-aa9c5a80-d01b-11eb-914f-5523cf80d5ba.png)


Basically I'm prototyping something that needs a pretty detailed point cloud with full object / VR overlay on top

**Comment by sebaFP** [comment]
> The polyfill seems to work quite ok. I've just hacked together an AR version on iOS. The orientation is not perfect, Euler and no roll, but ik looks ok: ![IMG_0302](https://user-images.githubusercontent.com/5172371/83353977-89a49100-a356-11ea-98a5-08e1b47c7e2c.gif)

Hi,
How do you achieve this?
Thanks!
